Vim: enable sending multiple keystrokes from custom keybinding

avi-cenna opened the pull request:
Release Notes:

- Added `workspace::SendKeystrokes` to enable mapping from one key to a sequence of others ([#7033](https://github.com/zed-industries/zed/issues/7033)). 

Improves #7033. Big thank you to @ConradIrwin who did most of the heavy lifting on this one.

This PR allows the user to send multiple keystrokes via custom keybinding. For example, the following keybinding would go down four lines and then right four characters.

```json
[
  {
    "context": "Editor && VimControl && !VimWaiting && !menu",
    "bindings": {
      "g z": [
        "workspace::SendKeystrokes",
        "j j j j l l l l"
      ],
    }
  }
]
```

avi-cenna committed:
wip

avi-cenna committed:
format

avi-cenna committed:
test

avi-cenna committed:
Revert "wip"

This reverts commit 1b753879

avi-cenna committed:
cleanup

ConradIrwin commented:
@avi-cenna Thanks for pushing this up!

#8019 fixes some of the issues we were seeing; but there are a number of cases we weren't thinking about on Friday that I'll add code for to this branch.

ConradIrwin committed:
Merge branch 'main' into vim/7033-send-multiple-keystrokes-part-1

ConradIrwin committed:
Allow dispatching keystrokes to the real app

This moves some of our test helper into the main window context to
power workspace::SendKeystrokes

avi-cenna commented on crates/workspace/src/workspace.rs:
```diff
@@ -1259,21 +1261,31 @@ impl Workspace {
-        })
+        keystrokes.reverse();
+
+        self.dispatching_keystrokes
```
What's the purpose of this `dispatching_keystrokes` field? It looks like the parsed keystrokes are appended to this field, but I don't see where the field is being used outside of this.

ConradIrwin replied:
It allows you to include keys that themselves are bound to `SendKeystrokes`, so you can call this recursively.

In order to do that reliably we need to make sure that we append the new keystrokes to the existing list to be sure they'll be processed in order.

avi-cenna commented:
Thanks @ConradIrwin . I was looking over the changes and just had some questions out of curiosity.

ConradIrwin commented on crates/workspace/src/workspace.rs:
```diff
@@ -1259,21 +1261,31 @@ impl Workspace {
-                cx.dispatch_event(PlatformInput::KeyUp(KeyUpEvent { keystroke }));
-            }
-        })
+        keystrokes.reverse();
```
Vectors in Rust are designed to be pushed/popped on the end; so this just makes the code a little more efficient. If you do `.remove(0)` to pull the first item out of the vector then it rewrites the entire vector, but `.pop()` can just change the length.

ConradIrwin committed:
Make command palette mostly synchronous
